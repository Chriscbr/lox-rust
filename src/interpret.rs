mod env;
mod value;

use std::{cell::RefCell, collections::HashMap, rc::Rc, sync::Arc};

use crate::ast::{
    Assign, Binary, BinaryOp, Call, Class, Expr, Function, Grouping, If, Literal, Logical,
    LogicalOp, Print, Return, Stmt, Unary, UnaryOp, VarDecl, Variable, While,
};

use self::{env::Environment, value::RuntimeValue};

pub struct Interpreter {
    env: Rc<RefCell<Environment>>,
    globals: Rc<RefCell<Environment>>,
    inside_function: bool,
}

impl Interpreter {
    pub fn new() -> Self {
        let globals = Rc::new(RefCell::new(Environment::new(None)));
        // globals.define(
        //     "clock",
        //     RuntimeValue::Callable(Callable {
        //         arity: 0,
        //         closure: value::clock,
        //     }),
        // );

        Self {
            env: globals.clone(),
            globals,
            inside_function: false,
        }
    }

    pub fn interpret(&mut self, stmts: &[Stmt]) -> Result<(), RuntimeError> {
        for stmt in stmts {
            self.execute(stmt)?;
        }
        Ok(())
    }

    fn execute(&mut self, stmt: &Stmt) -> Result<(), RuntimeError> {
        // dbg!(&stmt);
        // dbg!(&self.env);
        match stmt {
            Stmt::Block(stmts) => {
                let env = Rc::new(RefCell::new(Environment::new(Some(self.env.clone()))));
                self.execute_block(stmts, env)
            }
            Stmt::Class(class) => self.execute_class(class),
            Stmt::Expr(expr) => self.execute_expr(expr),
            Stmt::Function(func) => self.execute_function_decl(func),
            Stmt::If(i) => self.execute_if(i),
            Stmt::Print(p) => self.execute_print(p),
            Stmt::Return(r) => self.execute_return(r),
            Stmt::VarDecl(var_decl) => self.execute_var_decl(var_decl),
            Stmt::While(w) => self.execute_while(w),
        }?;
        Ok(())
    }

    fn execute_block(
        &mut self,
        stmts: &[Stmt],
        env: Rc<RefCell<Environment>>,
    ) -> Result<(), RuntimeError> {
        let previous = std::mem::replace(&mut self.env, env);
        for stmt in stmts {
            match self.execute(stmt) {
                Ok(_) => {}
                Err(RuntimeError::ReturnValue(v)) => {
                    self.env = previous;
                    return Err(RuntimeError::ReturnValue(v));
                }
                Err(e) => {
                    self.env = previous;
                    return Err(e);
                }
            }
        }
        self.env = previous;
        Ok(())
    }

    fn execute_class(&mut self, class: &Class) -> Result<(), RuntimeError> {
        let mut methods = HashMap::new();
        for method in &class.methods {
            let value = RuntimeValue::Function(value::Function::new(
                method.params.len(),
                method.clone(),
                self.env.clone(),
            ));
            methods.insert(method.name.clone(), value);
        }
        let class_value = RuntimeValue::Class(value::Class::new(class.name.clone(), methods));
        let new_env = self.env.borrow().extend(&class.name, class_value)?;
        self.env = Rc::new(RefCell::new(new_env));
        Ok(())
    }

    fn execute_expr(&mut self, expr: &Expr) -> Result<(), RuntimeError> {
        self.expr(expr)?;
        Ok(())
    }

    fn execute_function_decl(&mut self, fun: &Arc<Function>) -> Result<(), RuntimeError> {
        // let func = RuntimeValue::Function(value::Function::new(
        //     fun.params.len(),
        //     fun.clone(),
        //     self.env.clone(), // this will be replaced when the function is called
        // ));
        // let new_env = self.env.borrow().define(&fun.name, func)?;
        // self.env = Rc::new(RefCell::new(new_env));
        // self.env.borrow().assign(&fun.name, func)?;
        // let func = self.env.borrow().get(&fun.name)?;
        // match func {
        //     RuntimeValue::Function(mut f) => {
        //         f.closure = self.env.clone();
        //     }
        //     _ => unreachable!(),
        // }
        // Ok(())

        let func = RuntimeValue::Function(value::Function::new(
            fun.params.len(),
            fun.clone(),
            self.env.clone(),
        ));
        if self.env.borrow().is_global() {
            self.env.borrow_mut().set_global(&fun.name, func);
            return Ok(());
        }

        let new_env = self.env.borrow().extend(&fun.name, func)?;
        self.env = Rc::new(RefCell::new(new_env));
        let func = self.env.borrow().get(&fun.name)?;
        match func {
            RuntimeValue::Function(mut f) => {
                f.closure = self.env.clone();
            }
            _ => unreachable!(),
        }
        Ok(())

        // let temp = RuntimeValue::Nil;
        // let new_env = self.env.borrow().extend(&fun.name, temp)?;
        // self.env = Rc::new(RefCell::new(new_env));
        // let func = RuntimeValue::Function(value::Function::new(
        //     fun.params.len(),
        //     fun.clone(),
        //     self.env.clone(),
        // ));
        // self.env.borrow().assign(&fun.name, func)?;
        // let func = self.env.borrow().get(&fun.name)?;
        // match func {
        //     RuntimeValue::Function(mut f) => {
        //         f.closure = self.env.clone();
        //     }
        //     _ => unreachable!(),
        // }
        // Ok(())
    }

    fn execute_if(&mut self, i: &If) -> Result<(), RuntimeError> {
        let cond = self.expr(&i.condition)?;
        if self.is_truthy(&cond) {
            self.execute(&i.then_branch)?;
        } else if let Some(else_branch) = &i.else_branch {
            self.execute(else_branch)?;
        }
        Ok(())
    }

    fn execute_print(&mut self, p: &Print) -> Result<(), RuntimeError> {
        let value = self.expr(&p.expr)?;
        println!("{}", value);
        Ok(())
    }

    fn execute_return(&mut self, r: &Return) -> Result<(), RuntimeError> {
        if !self.inside_function {
            return Err(RuntimeError::CannotReturnFromTopLevel);
        }
        let value = match &r.value {
            Some(expr) => self.expr(expr)?,
            None => RuntimeValue::Nil,
        };
        Err(RuntimeError::ReturnValue(value))
    }

    fn execute_var_decl(&mut self, var_decl: &VarDecl) -> Result<(), RuntimeError> {
        let value = match &var_decl.initializer {
            Some(expr) => self.expr(expr)?,
            None => RuntimeValue::Nil,
        };
        if self.env.borrow().is_global() {
            self.env.borrow_mut().set_global(&var_decl.name, value);
            return Ok(());
        }
        let new_env = self.env.borrow().extend(&var_decl.name, value)?;
        self.env = Rc::new(RefCell::new(new_env));
        Ok(())
    }

    fn execute_while(&mut self, w: &While) -> Result<(), RuntimeError> {
        let mut cond = self.expr(&w.condition)?;
        while self.is_truthy(&cond) {
            self.execute(&w.body)?;
            cond = self.expr(&w.condition)?;
        }
        Ok(())
    }

    fn expr(&mut self, expr: &Expr) -> Result<RuntimeValue, RuntimeError> {
        match expr {
            Expr::Binary(b) => self.binary(b),
            Expr::Call(c) => self.call(c),
            Expr::Grouping(g) => self.grouping(g),
            Expr::Literal(l) => self.literal(l),
            Expr::Logical(l) => self.logical(l),
            Expr::Unary(u) => self.unary(u),
            Expr::Variable(v) => self.variable(v),
            Expr::Assign(a) => self.assign(a),
        }
    }

    fn binary(&mut self, binary: &Binary) -> Result<RuntimeValue, RuntimeError> {
        let left = self.expr(&binary.left)?;
        let right = self.expr(&binary.right)?;

        match binary.op {
            BinaryOp::Add => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Number(l + r))
                }
                (RuntimeValue::String(l), RuntimeValue::String(r)) => {
                    Ok(RuntimeValue::String(format!("{}{}", l, r)))
                }
                _ => Err(RuntimeError::OperandsMustBeTwoNumbersOrTwoStrings),
            },
            BinaryOp::Sub => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Number(l - r))
                }
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::Mul => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Number(l * r))
                }
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::Div => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Number(l / r))
                }
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::Eq => Ok(RuntimeValue::Bool(self.is_equal(&left, &right))),
            BinaryOp::NotEq => Ok(RuntimeValue::Bool(!self.is_equal(&left, &right))),
            BinaryOp::Lt => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => Ok(RuntimeValue::Bool(l < r)),
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::LtEq => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Bool(l <= r))
                }
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::Gt => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => Ok(RuntimeValue::Bool(l > r)),
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
            BinaryOp::GtEq => match (left, right) {
                (RuntimeValue::Number(l), RuntimeValue::Number(r)) => {
                    Ok(RuntimeValue::Bool(l >= r))
                }
                _ => Err(RuntimeError::OperandsMustBeNumbers),
            },
        }
    }

    fn call(&mut self, call: &Call) -> Result<RuntimeValue, RuntimeError> {
        let callee = self.expr(&call.callee)?;

        let mut args = Vec::new();
        for arg in &call.args {
            args.push(self.expr(arg)?);
        }

        match callee {
            RuntimeValue::Function(f) => f.call(self, args),
            _ => Err(RuntimeError::NotCallable),
        }
    }

    fn grouping(&mut self, grouping: &Grouping) -> Result<RuntimeValue, RuntimeError> {
        self.expr(&grouping.expr)
    }

    fn literal(&self, literal: &Literal) -> Result<RuntimeValue, RuntimeError> {
        match literal {
            Literal::Number(n) => Ok(RuntimeValue::Number(*n)),
            Literal::String(s) => Ok(RuntimeValue::String(s.clone())),
            Literal::Bool(b) => Ok(RuntimeValue::Bool(*b)),
            Literal::Nil => Ok(RuntimeValue::Nil),
        }
    }

    fn logical(&mut self, logical: &Logical) -> Result<RuntimeValue, RuntimeError> {
        let left = self.expr(&logical.left)?;

        match logical.op {
            LogicalOp::And => {
                if !self.is_truthy(&left) {
                    return Ok(left);
                }
            }
            LogicalOp::Or => {
                if self.is_truthy(&left) {
                    return Ok(left);
                }
            }
        }

        self.expr(&logical.right)
    }

    fn unary(&mut self, unary: &Unary) -> Result<RuntimeValue, RuntimeError> {
        let right = self.expr(&unary.right)?;

        match unary.op {
            UnaryOp::Negate => match right {
                RuntimeValue::Number(n) => Ok(RuntimeValue::Number(-n)),
                _ => Err(RuntimeError::OperandMustBeNumber),
            },
            UnaryOp::Not => Ok(RuntimeValue::Bool(!self.is_truthy(&right))),
        }
    }

    fn variable(&self, variable: &Variable) -> Result<RuntimeValue, RuntimeError> {
        self.lookup_variable(&variable.name)
    }

    fn assign(&mut self, assign: &Assign) -> Result<RuntimeValue, RuntimeError> {
        let value = self.expr(&assign.value)?;
        self.assign_variable(&assign.name, value.clone())?;
        Ok(value)
    }

    fn lookup_variable(&self, name: &str) -> Result<RuntimeValue, RuntimeError> {
        self.env.borrow().get(name)
        // if let Ok(value) = self.env.borrow().get(name) {
        //     Ok(value)
        // } else {
        //     self.globals.borrow().get(name)
        // }
    }

    fn assign_variable(&mut self, name: &str, value: RuntimeValue) -> Result<(), RuntimeError> {
        if self.env.borrow().is_global() {
            self.env.borrow_mut().set_global(name, value);
            return Ok(());
        }
        self.env.borrow().assign(name, value)
    }

    fn is_truthy(&self, value: &RuntimeValue) -> bool {
        match value {
            RuntimeValue::Nil => false,
            RuntimeValue::Bool(b) => *b,
            _ => true,
        }
    }

    fn is_equal(&self, a: &RuntimeValue, b: &RuntimeValue) -> bool {
        match (a, b) {
            (RuntimeValue::Nil, RuntimeValue::Nil) => true,
            (RuntimeValue::Number(l), RuntimeValue::Number(r)) => l == r,
            (RuntimeValue::String(l), RuntimeValue::String(r)) => l == r,
            (RuntimeValue::Bool(l), RuntimeValue::Bool(r)) => l == r,
            _ => false,
        }
    }
}

#[derive(Debug)]
pub enum RuntimeError {
    OperandsMustBeNumbers,
    OperandsMustBeTwoNumbersOrTwoStrings,
    OperandMustBeNumber,
    UndefinedVariable(String),
    InvalidArgumentCount,
    NotCallable,
    AlreadyDefined(String),
    CannotReturnFromTopLevel,
    ReturnValue(RuntimeValue),
}

impl std::fmt::Display for RuntimeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RuntimeError::OperandsMustBeNumbers => write!(f, "Operands must be numbers."),
            RuntimeError::OperandsMustBeTwoNumbersOrTwoStrings => {
                write!(f, "Operands must be two numbers or two strings.")
            }
            RuntimeError::OperandMustBeNumber => write!(f, "Operand must be a number."),
            RuntimeError::UndefinedVariable(name) => write!(f, "Undefined variable '{}'.", name),
            RuntimeError::InvalidArgumentCount => write!(f, "Invalid argument count."),
            RuntimeError::NotCallable => write!(f, "Can only call functions and classes."),
            RuntimeError::AlreadyDefined(name) => {
                write!(
                    f,
                    "Already a variable with the name '{}' in this scope.",
                    name
                )
            }
            RuntimeError::CannotReturnFromTopLevel => {
                write!(f, "Cannot return from top-level code.")
            }
            RuntimeError::ReturnValue(_) => panic!("Internal error: unhandled return value"),
        }
    }
}
