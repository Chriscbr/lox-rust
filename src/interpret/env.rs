use std::collections::HashMap;

use super::{value::RuntimeValue, RuntimeError};

pub struct Environment {
    values: HashMap<String, RuntimeValue>,
}

impl Environment {
    pub fn new() -> Self {
        Self {
            values: HashMap::new(),
        }
    }

    pub fn define(&mut self, name: &str, value: RuntimeValue) {
        self.values.insert(name.to_string(), value);
    }

    pub fn get(&self, name: &str) -> Result<RuntimeValue, RuntimeError> {
        match self.values.get(name) {
            Some(value) => Ok(value.clone()),
            None => Err(RuntimeError::UndefinedVariable(name.to_string())),
        }
    }
}
