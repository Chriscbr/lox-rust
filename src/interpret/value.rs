use std::{cell::RefCell, collections::HashMap, fmt::Display, rc::Rc, sync::Arc};

use super::{env::Environment, Interpreter, RuntimeError};

use crate::ast;

#[derive(Debug, Clone)]
pub enum RuntimeValue {
    Nil,
    Bool(bool),
    Number(f64),
    String(String),
    Function(Function),
    Class(Class),
    Instance(Instance),
}

#[derive(Debug, Clone)]
pub struct Function {
    pub arity: usize,
    pub fun: Arc<ast::Function>,
    pub closure: Rc<RefCell<Environment>>,
}

impl Function {
    pub fn new(arity: usize, fun: Arc<ast::Function>, closure: Rc<RefCell<Environment>>) -> Self {
        Self {
            arity,
            fun,
            closure,
        }
    }

    pub fn call(
        &self,
        interpreter: &mut Interpreter,
        args: Vec<RuntimeValue>,
    ) -> Result<RuntimeValue, RuntimeError> {
        if args.len() != self.arity {
            return Err(RuntimeError::InvalidArgumentCount);
        }

        let env = Rc::new(RefCell::new(Environment::new(Some(self.closure.clone()))));
        for (i, param) in self.fun.params.iter().enumerate() {
            env.borrow_mut().define(param, args[i].clone())?;
        }

        let prev_inside_function = interpreter.inside_function;
        interpreter.inside_function = true;
        let result = interpreter.execute_block(&self.fun.body, env);
        interpreter.inside_function = prev_inside_function;

        match result {
            Ok(_) => Ok(RuntimeValue::Nil),
            Err(RuntimeError::ReturnValue(value)) => Ok(value),
            Err(e) => Err(e),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Class {
    pub name: String,
    pub methods: HashMap<String, RuntimeValue>,
}

impl Class {
    pub fn new(name: String, methods: HashMap<String, RuntimeValue>) -> Self {
        Self { name, methods }
    }
}

#[derive(Debug, Clone)]
pub struct Instance {
    pub class: Arc<Class>,
    pub fields: HashMap<String, RuntimeValue>,
}

impl Instance {
    pub fn new(class: Arc<Class>) -> Self {
        Self {
            class,
            fields: HashMap::new(),
        }
    }
}

impl Display for RuntimeValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            RuntimeValue::Nil => format!("nil"),
            RuntimeValue::Bool(b) => format!("{}", b),
            RuntimeValue::Number(n) => {
                let text = format!("{}", n);
                if text.ends_with(".0") {
                    text[..text.len() - 2].to_string()
                } else {
                    text
                }
            }
            RuntimeValue::String(s) => format!("{}", s),
            RuntimeValue::Function(c) => format!("<fn {}>", c.fun.name),
            RuntimeValue::Class(c) => format!("{}", c.name),
            RuntimeValue::Instance(i) => format!("<instance of {}>", i.class.name),
        };
        write!(f, "{}", s)
    }
}

pub fn clock(_: &[RuntimeValue]) -> Result<RuntimeValue, RuntimeError> {
    Ok(RuntimeValue::Number(
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64(),
    ))
}
