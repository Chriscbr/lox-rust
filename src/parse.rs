use crate::{
    ast::{Binary, BinaryOp, Expr, Grouping, Literal, Stmt, Unary, UnaryOp, VarDecl, Variable},
    error::Error,
    token::{Token, TokenType},
};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
    errors: Vec<Error>,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            current: 0,
            errors: vec![],
        }
    }

    pub fn parse(&mut self) -> Result<Vec<Stmt>, Error> {
        let mut stmts = vec![];
        while !self.is_at_end() {
            let stmt = self.declaration();
            match stmt {
                Ok(stmt) => stmts.push(stmt),
                Err(err) => {
                    self.errors.push(err);
                    self.synchronize();
                }
            }
        }

        Ok(stmts)
    }

    fn expr(&mut self) -> Result<Expr, Error> {
        self.equality()
    }

    fn declaration(&mut self) -> Result<Stmt, Error> {
        if self.matches(&[TokenType::Var]) {
            return self.var_declaration();
        }

        self.stmt()
    }

    fn stmt(&mut self) -> Result<Stmt, Error> {
        if self.matches(&[TokenType::Print]) {
            return self.print_statement();
        }

        self.expr_stmt()
    }

    fn print_statement(&mut self) -> Result<Stmt, Error> {
        let expr = self.expr()?;
        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::Print(expr))
    }

    fn var_declaration(&mut self) -> Result<Stmt, Error> {
        if !self.check(TokenType::Identifier) {
            return Err(Error::ExpectedVariableName {
                token: self.peek().clone(),
            });
        }
        let name = self.advance().lexeme.clone();
        let initializer = if self.matches(&[TokenType::Equal]) {
            Some(self.expr()?)
        } else {
            None
        };

        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::VarDecl(VarDecl { name, initializer }))
    }

    fn expr_stmt(&mut self) -> Result<Stmt, Error> {
        let expr = self.expr()?;
        if !self.check(TokenType::Semicolon) {
            return Err(Error::ExpectSemicolonAfterExpression {
                token: self.peek().clone(),
            });
        }
        self.advance();
        Ok(Stmt::Expr(expr))
    }

    fn equality(&mut self) -> Result<Expr, Error> {
        let mut expr = self.comparison()?;

        while self.matches(&[TokenType::BangEqual, TokenType::EqualEqual]) {
            let op = self.previous().ty;
            let right = self.comparison()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn comparison(&mut self) -> Result<Expr, Error> {
        let mut expr = self.term()?;

        while self.matches(&[
            TokenType::Greater,
            TokenType::GreaterEqual,
            TokenType::Less,
            TokenType::LessEqual,
        ]) {
            let op = self.previous().ty;
            let right = self.term()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn term(&mut self) -> Result<Expr, Error> {
        let mut expr = self.factor()?;

        while self.matches(&[TokenType::Minus, TokenType::Plus]) {
            let op = self.previous().ty;
            let right = self.factor()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn factor(&mut self) -> Result<Expr, Error> {
        let mut expr = self.unary()?;

        while self.matches(&[TokenType::Slash, TokenType::Star]) {
            let op = self.previous().ty;
            let right = self.unary()?;
            expr = Expr::Binary(Binary {
                left: Box::new(expr),
                op: BinaryOp::from(op),
                right: Box::new(right),
            });
        }

        Ok(expr)
    }

    fn unary(&mut self) -> Result<Expr, Error> {
        if self.matches(&[TokenType::Bang, TokenType::Minus]) {
            let op = self.previous().ty;
            let right = self.unary()?;
            return Ok(Expr::Unary(Unary {
                op: UnaryOp::from(op),
                right: Box::new(right),
            }));
        }

        return self.primary();
    }

    fn primary(&mut self) -> Result<Expr, Error> {
        if self.matches(&[TokenType::False]) {
            return Ok(Expr::Literal(Literal::Bool(false)));
        }

        if self.matches(&[TokenType::True]) {
            return Ok(Expr::Literal(Literal::Bool(true)));
        }

        if self.matches(&[TokenType::Nil]) {
            return Ok(Expr::Literal(Literal::Nil));
        }

        if self.matches(&[TokenType::Number]) {
            let value = str::parse::<f64>(&self.previous().lexeme).expect("invalid float");
            return Ok(Expr::Literal(Literal::Number(value)));
        }

        if self.matches(&[TokenType::String]) {
            let s = &self.previous().lexeme;
            // Trim the surrounding quotes.
            let trimmed = &s[1..s.len() - 1];
            return Ok(Expr::Literal(Literal::String(trimmed.to_string())));
        }

        if self.matches(&[TokenType::Identifier]) {
            return Ok(Expr::Variable(Variable {
                name: self.previous().lexeme.clone(),
            }));
        }

        if self.matches(&[TokenType::LeftParen]) {
            let expr = self.expr()?;
            if !self.check(TokenType::RightParen) {
                return Err(Error::ExpectRightParenAfterExpr {
                    token: self.peek().clone(),
                });
            }
            self.advance();
            return Ok(Expr::Grouping(Grouping {
                expr: Box::new(expr),
            }));
        }

        todo!("expected expression");
    }

    fn matches(&mut self, tys: &[TokenType]) -> bool {
        for ty in tys {
            if self.check(*ty) {
                self.advance();
                return true;
            }
        }

        false
    }

    fn check(&self, ty: TokenType) -> bool {
        if self.is_at_end() {
            return false;
        }

        return self.peek().ty == ty;
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.current += 1;
        }

        return self.previous();
    }

    fn is_at_end(&self) -> bool {
        matches!(self.peek().ty, TokenType::EOF)
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }

    fn synchronize(&mut self) {
        while !self.is_at_end() {
            if self.previous().ty == TokenType::Semicolon {
                return;
            }

            match self.peek().ty {
                TokenType::Class
                | TokenType::Fun
                | TokenType::Var
                | TokenType::For
                | TokenType::If
                | TokenType::While
                | TokenType::Print
                | TokenType::Return => return,
                _ => {}
            }

            self.advance();
        }
    }
}
