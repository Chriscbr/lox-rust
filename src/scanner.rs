use crate::{
    error::Error,
    token::{Token, TokenType, KEYWORDS},
};

#[derive(Debug)]
pub struct Scanner {
    source: String, // chars: Vec<char> ?
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
    errors: Vec<Error>,
}

impl Scanner {
    pub fn new(source: String) -> Self {
        Self {
            source,
            tokens: vec![],
            start: 0,
            current: 0,
            line: 1,
            errors: vec![],
        }
    }

    pub fn scan_tokens(mut self) -> (Vec<Token>, Vec<Error>) {
        while !self.is_at_end() {
            // We are at the beginning of the next lexeme.
            self.start = self.current;
            self.scan_token();
        }

        self.tokens
            .push(Token::new(TokenType::EOF, "".to_string(), self.line));
        (self.tokens, self.errors)
    }

    fn is_at_end(&self) -> bool {
        return self.current >= self.source.len();
    }

    fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '(' => self.add_token(TokenType::LeftParen),
            ')' => self.add_token(TokenType::RightParen),
            '{' => self.add_token(TokenType::LeftBrace),
            '}' => self.add_token(TokenType::RightBrace),
            ',' => self.add_token(TokenType::Comma),
            '.' => self.add_token(TokenType::Dot),
            '-' => self.add_token(TokenType::Minus),
            '+' => self.add_token(TokenType::Plus),
            ';' => self.add_token(TokenType::Semicolon),
            '*' => self.add_token(TokenType::Star),
            '!' => {
                let ty = if self.matches('=') {
                    TokenType::BangEqual
                } else {
                    TokenType::Bang
                };
                self.add_token(ty);
            }
            '=' => {
                let ty = if self.matches('=') {
                    TokenType::EqualEqual
                } else {
                    TokenType::Equal
                };
                self.add_token(ty);
            }
            '<' => {
                let ty = if self.matches('=') {
                    TokenType::LessEqual
                } else {
                    TokenType::Less
                };
                self.add_token(ty);
            }
            '>' => {
                let ty = if self.matches('=') {
                    TokenType::GreaterEqual
                } else {
                    TokenType::Greater
                };
                self.add_token(ty);
            }
            '/' => {
                if self.matches('/') {
                    // A comment goes until the end of the line.
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else {
                    self.add_token(TokenType::Slash);
                }
            }
            ' ' | '\r' | '\t' => {} // Ignore whitespace.
            '\n' => {
                self.line += 1;
            }
            '"' => {
                self.string();
            }
            _ => {
                if self.is_digit(c) {
                    self.number();
                } else if self.is_alpha(c) {
                    self.identifier();
                } else {
                    self.errors
                        .push(Error::UnexpectedCharacter { line: self.line });
                }
            }
        };
    }

    fn identifier(&mut self) {
        while self.is_alpha_numeric(self.peek()) {
            self.advance();
        }

        let text = &self.source[self.start..self.current];
        let ty = KEYWORDS.get(text);
        match ty {
            Some(ty) => self.add_token(*ty),
            None => self.add_token(TokenType::Identifier),
        }

        self.add_token(TokenType::Identifier)
    }

    fn number(&mut self) {
        while self.is_digit(self.peek()) {
            self.advance();
        }

        // Look for a fractional part.
        if self.peek() == '.' && self.is_digit(self.peek_next()) {
            // Consume the "."
            self.advance();

            while self.is_digit(self.peek()) {
                self.advance();
            }
        }

        self.add_token(TokenType::Number);
    }

    fn string(&mut self) {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            self.errors
                .push(Error::UnterminatedString { line: self.line });
        }

        self.advance(); // The closing ".

        self.add_token(TokenType::String);
    }

    fn matches(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.char_at(self.current) != expected {
            return false;
        }

        self.current += 1;
        return true;
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        return self.char_at(self.current);
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }
        return self.char_at(self.current + 1);
    }

    fn is_alpha(&self, c: char) -> bool {
        c.is_ascii_alphabetic() || c == '_'
    }

    fn is_alpha_numeric(&self, c: char) -> bool {
        self.is_alpha(c) || self.is_digit(c)
    }

    fn is_digit(&self, c: char) -> bool {
        c.is_ascii_digit()
    }

    fn advance(&mut self) -> char {
        let c = self.char_at(self.current);
        self.current += 1;
        c
    }

    fn add_token(&mut self, ty: TokenType) {
        let text = &self.source[self.start..self.current];
        self.tokens
            .push(Token::new(ty, text.to_string(), self.line)); // str -> String
    }

    fn char_at(&self, idx: usize) -> char {
        return char::from(self.source.as_bytes()[idx]); // unsafe conversion
    }
}
