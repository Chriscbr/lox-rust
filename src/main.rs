mod ast;
mod interpret;
mod parse;
mod scanner;
mod token;

use std::{fs::read_to_string, io::Write};

use clap::Parser as ClapParser;
use interpret::Interpreter;
use parse::{report_parse_errors, Parser};
use scanner::Scanner;

/// Run a Lox program.
#[derive(ClapParser, Debug)]
#[command(version, about)]
struct Args {
    /// File path
    file: String,
}

fn main() {
    let args = Args::parse();
    let source_text = read_to_string(args.file).expect("unable to read file");

    let mut stdout = std::io::stdout();
    let mut stderr = std::io::stderr();
    run(source_text, &mut stdout, &mut stderr);
}

fn run<W: Write, X: Write>(source: String, stdout: &mut W, stderr: &mut X) {
    let scanner = Scanner::new(source);
    let (tokens, errors) = scanner.scan_tokens();

    if errors.len() > 0 {
        for (error, line) in errors {
            writeln!(stderr, "[line {}] Error: {}", line, error).unwrap();
        }
        return;
    }

    let parser = Parser::new(tokens);
    let parsed = parser.parse();

    let stmts = match parsed {
        Ok(expr) => expr,
        Err(errors) => {
            report_parse_errors(errors, stderr);
            return;
        }
    };

    let wrapper = WriteWrapper { writer: stdout };
    let boxed_writer = Box::new(wrapper);
    let mut interpreter = Interpreter::new(boxed_writer);
    if let Err(err) = interpreter.interpret(&stmts) {
        writeln!(stderr, "{}", err).unwrap();
    }
}

struct WriteWrapper<'a, W: Write + ?Sized> {
    writer: &'a mut W,
}

impl<'a, W: Write + ?Sized> Write for WriteWrapper<'a, W> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.writer.write(buf)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.writer.flush()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_examples() {
        let examples_dir = std::fs::read_dir("examples").unwrap();

        for example in examples_dir {
            eprintln!("{:?}", example);
            let example = example.unwrap();
            let file_name = example.file_name();
            let file_name = file_name.to_str().unwrap();
            let source_text = read_to_string(format!("examples/{}", file_name)).unwrap();

            let mut stdout = Vec::new();
            let mut stderr = Vec::new();
            run(source_text, &mut stdout, &mut stderr);

            let contents = format!(
                "stdout:\n{}\nstderr:\n{}",
                String::from_utf8(stdout).unwrap(),
                String::from_utf8(stderr).unwrap()
            );

            insta::assert_snapshot!(file_name, contents);
        }
    }
}
