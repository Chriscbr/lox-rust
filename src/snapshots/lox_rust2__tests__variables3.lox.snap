---
source: src/main.rs
expression: contents
---
stdout:

stderr:
Undefined variable 'a'.
